src/main/java/com/uid2/shared/secret/SecureKeyGenerator.java

```diff
@@ -22,11 +28,57 @@ public String generateRandomKeyString(int keyLen) {
     }
 
     @Override
-    public String generateFormattedKeyString(int keyLen) {
+    public KeyGenerationResult generateFormattedKeyStringAndKeyHash(String keyPrefix, int keyLen) {
         String key = this.generateRandomKeyString(keyLen);
-        if (key.length() < 6) {
-            return key;
+        String formattedKey = key.length() >= 6 ? new StringBuilder(key).insert(6, ".").toString() : key;
+
+        byte[] keyBytes = Utils.decodeBase64String(key);
+        byte[] saltBytes = generateRandomKey(keyLen);
+
+        MessageDigest md = createMessageDigest();
+        md.update(saltBytes);
+        byte[] hashBytes = md.digest(keyBytes); // This will always generate a byte array of length 512 (64 bytes)
+
+        String hash = Utils.toBase64String(hashBytes); // This will always convert the hashBytes to a String with 88 chars (86 + 2 padding)
+        String salt = Utils.toBase64String(saltBytes);
+        String keyHash = String.format("%s$%s", hash, salt);
+
+        return new KeyGenerationResult(keyPrefix + formattedKey, keyPrefix + keyHash);
+    }
+
+    @Override
+    public boolean compareFormattedKeyStringAndKeyHash(String formattedInputKey, String keyHash) {

```

Comment:
This shouldn't exist either, let's keep the hashing as simple as possible. Mostly to support the old key format as well